Tests/SonarScanner.MSBuild.PreProcessor.Test/ProcessedArgsTests.cs

```diff
@@ -361,12 +390,86 @@ public void ProcArgs_OperatingSystem_AutoDetection(PlatformOS platformOS, string
             operatingSystemProvider.OperatingSystem().Returns(_ => platformOS);
             var sut = CreateDefaultArgs(operatingSystemProvider: operatingSystemProvider);
             sut.OperatingSystem.Should().Be(expectedOperatingSystem);
+            sut.IsValid.Should().BeTrue();
+        }
+
+        [TestMethod]
+        public void ProcArgs_UserHome_ParameterProvided()
+        {
+            var directoryWrapper = Substitute.For<IDirectoryWrapper>();
+            directoryWrapper.Exists(@"C:\Users\user\.sonar").Returns(true);
+            var sut = CreateDefaultArgs(new ListPropertiesProvider([new Property(SonarProperties.UserHome, @"C:\Users\user\.sonar")]), directoryWrapper: directoryWrapper);
+            sut.UserHome.Should().Be(@"C:\Users\user\.sonar");
+            sut.IsValid.Should().BeTrue();
+            logger.AssertNoErrorsLogged();
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public void ProcArgs_UserHome_ParameterProvided_DoesNotExists()
+        {
+            var directoryWrapper = Substitute.For<IDirectoryWrapper>();
+            directoryWrapper.Exists(@"C:\Users\user\.sonar").Returns(false);
+            var sut = CreateDefaultArgs(new ListPropertiesProvider([new Property(SonarProperties.UserHome, @"C:\Users\user\.sonar")]), directoryWrapper: directoryWrapper);
+            sut.UserHome.Should().BeNull();
+            sut.IsValid.Should().BeFalse();
+            logger.AssertErrorLogged(@"The provided value for 'sonar.userHome' 'C:\Users\user\.sonar' does not exists. Specify a valid directory for 'sonar.userHome'.");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public void ProcArgs_UserHome_Default()
+        {
+            var directoryWrapper = Substitute.For<IDirectoryWrapper>();
+            var operatingSystemProvider = Substitute.For<IOperatingSystemProvider>();
+            operatingSystemProvider.GetFolderPath(Environment.SpecialFolder.UserProfile, Environment.SpecialFolderOption.None).Returns(@"C:\Users\user");
+            directoryWrapper.Exists(@"C:\Users\user\.sonar").Returns(true);
+            var sut = CreateDefaultArgs(directoryWrapper: directoryWrapper, operatingSystemProvider: operatingSystemProvider);
+            sut.UserHome.Should().Be(@"C:\Users\user\.sonar");
+            sut.IsValid.Should().BeTrue();
+            logger.AssertNoErrorsLogged();
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public void ProcArgs_UserHome_Default_CreatedOnDemand()
+        {
+            var directoryWrapper = Substitute.For<IDirectoryWrapper>();
+            var operatingSystemProvider = Substitute.For<IOperatingSystemProvider>();
+            operatingSystemProvider.GetFolderPath(Environment.SpecialFolder.UserProfile, Environment.SpecialFolderOption.None).Returns(@"C:\Users\user");
```

Comment:
There is a bit of duplication with setting up the directory wrapper and the os provider. What do you think about extracting a couple of helper methods?

Reply:
I don't mind repetition in tests too much. For me, this is accidental duplication: These tests set up the environment the same way, but the environments are identical by accident. More so the setup of, e.g., `directoryWrapper.Exists(@"C:\Users\user\.sonar").Returns(false);` isn't even needed technically but emphasizes an important part the UT wants to test. If you move this logic to another place, the whole setup is hidden in some place far away from the UT making it hard to understand. Also, we need both directoryWrapper and operatingSystemProvider later in the assertion phase to make sure, e.g., that `CreateDirectory` was called.

Reply:
The extracted methods could return the substitutes, so assertions should not be a problem. At the same time, I understand your point of view about having the setup in the test. As long as we don't have to make changes very often, I think it's fine.